src/net_doh.c

```diff
@@ -0,0 +1,1140 @@
+    }
+
+    // guard against re-entrant http2_init
+    if (!self->http2->settings_sent) {
+        // send settings
+        ret = _http2_send_settings(self->http2);
+        if (ret != 0) {
+            perf_log_printf("nghttp2_submit_settings failed: %s", nghttp2_strerror(ret));
+            PERF_UNLOCK(&self->lock);
+            return 0;
+        }
+
+        self->http2->settings_sent = true;
+        // once we have TLS + http2 set, then we are ready to operate
+    }
+    
+    PERF_UNLOCK(&self->lock);
+
+    if (sock->event) {
+        sock->event(sock, self->conn_event, perf_get_time() - self->conn_ts);
+        self->conn_event = perf_socket_event_reconnected;
+    }
+
+    return 1;
+}
+
+static bool perf__doh_have_more(struct perf_net_socket* sock)
+{
+    return self->have_more;
+}
+
+#ifndef OPENSSL_NO_NEXTPROTONEG
+/* NPN TLS extension check */
+static int select_next_proto_cb(SSL *ssl, unsigned char **out,
+                                unsigned char *outlen, const unsigned char *in,
+                                unsigned int inlen, void *arg) {
+  (void)ssl;
+  (void)arg;
+
+  if (nghttp2_select_next_protocol(out, outlen, in, inlen) <= 0) {
+    perf_log_fatal("Server did not advertise %u", NGHTTP2_PROTO_VERSION_ID);
+  }
+
+  return SSL_TLSEXT_ERR_OK;
+}
+#endif /* !OPENSSL_NO_NEXTPROTONEG */
+
+struct perf_net_socket* perf_net_doh_opensocket(const perf_sockaddr_t* server, const perf_sockaddr_t* local, size_t bufsize)
+{
+    struct perf__doh_socket* tmp  = calloc(1, sizeof(struct perf__doh_socket)); // clang scan-build
+    struct perf_net_socket*  sock = (struct perf_net_socket*)tmp;
+
+    int ret = -1;
+
+    if (!sock) {
+        perf_log_fatal("perf_net_doh_opensocket() out of memory");
+        return 0; // needed for clang scan build
+    }
+
+    sock->recv      = perf__doh_recv;
+    sock->sendto    = perf__doh_sendto;
+    sock->close     = perf__doh_close;
+    sock->sockeq    = perf__doh_sockeq;
+    sock->sockready = perf__doh_sockready;
+    sock->have_more = perf__doh_have_more;
+
+    self->server  = *server;
+    self->local   = *local;
+    self->bufsize = bufsize;
+    if (self->bufsize > 0) {
+        self->bufsize *= 1024;
+    }
+    self->conning_event = perf_socket_event_connecting;
+    self->conn_event    = perf_socket_event_connected;
+    PERF_MUTEX_INIT(&self->lock);
+
+    if (!ssl_ctx) {
+#ifdef HAVE_TLS_METHOD
+        if (!(ssl_ctx = SSL_CTX_new(TLS_method()))) {
+            perf_log_fatal("SSL_CTX_new(): %s", ERR_error_string(ERR_get_error(), 0));
+        }
+        if (!SSL_CTX_set_min_proto_version(ssl_ctx, TLS1_2_VERSION)) {
+            perf_log_fatal("SSL_CTX_set_min_proto_version(TLS1_2_VERSION): %s", ERR_error_string(ERR_get_error(), 0));
+        }
+#else
+        perf_log_fatal("DNS-over-HTTPS (DoH) is supported only over TLS 1.2+");
+#endif
+        SSL_CTX_set_mode(ssl_ctx, SSL_MODE_ENABLE_PARTIAL_WRITE);
+        SSL_CTX_set_options(ssl_ctx,
```

Comment:
Why all these options?

Reply:
with the idea to enforce TLS 1.2 only as per https://datatracker.ietf.org/doc/html/rfc7540#section-9.2

Reply:
This is a testing tool, we need to be able to test anything. If restrictions like this should exist they need to be configurable.

I also don't understand what no compression has to do with this?

Reply:
okidoki, will drop the TLS1.2 requirement to leave it open to test any endpoint even if not http2 complaint as per RFC. 

NO_COMPRESSION: the idea was to minimise cpu bound overhead, will drop.

NO_SESSION_RESUMPTION_ON_RENEGOTIATION: trying to prevent renegotiation on established sessions as this will skew the performance results, thinking about it it should be probably OP_NO_RENEGOTIATION. Will drop it too.